File: server/routes/qr.py
"""QR code generation endpoint.

Returns a QR code image (PNG) that encodes the frontend map page URL.
Used at the hackathon so ~30 users can scan and join instantly.
"""

from __future__ import annotations

import io
import os

import qrcode
from fastapi import APIRouter, Query
from fastapi.responses import StreamingResponse

router = APIRouter(prefix="/api", tags=["qr"])

# Default frontend URL — overridable via FRONTEND_URL env var
_DEFAULT_FRONTEND_URL = "https://bulpan.example.com"


def _get_map_url(demo: bool = False) -> str:
    """Build the map page URL, optionally with ?demo=true for GPS fallback."""
    base = os.getenv("FRONTEND_URL", _DEFAULT_FRONTEND_URL)
    # Strip trailing slash for consistency
    base = base.rstrip("/")
    if demo:
        return f"{base}?demo=true"
    return base


def _generate_qr_png(data: str, box_size: int = 10, border: int = 2) -> bytes:
    """Generate a QR code PNG image as bytes."""
    qr = qrcode.QRCode(
        version=None,  # auto-size
        error_correction=qrcode.constants.ERROR_CORRECT_M,
        box_size=box_size,
        border=border,
    )
    qr.add_data(data)
    qr.make(fit=True)

    img = qr.make_image(fill_color="black", back_color="white")
    buf = io.BytesIO()
    img.save(buf, format="PNG")
    buf.seek(0)
    return buf.getvalue()


@router.get(
    "/qr",
    summary="Generate QR code for map page",
    responses={
        200: {
            "content": {"image/png": {}},
            "description": "QR code PNG image linking to the map page",
        }
    },
)
async def generate_qr(
    demo: bool = Query(False, description="Include ?demo=true for GPS fallback"),
    size: int = Query(10, ge=4, le=40, description="QR box size in pixels (4-40)"),
):
    """Return a QR code image (PNG) that links to the frontend map page.

    Scan this QR at the hackathon venue to instantly open the fire map.
    """
    url = _get_map_url(demo=demo)
    png_bytes = _generate_qr_png(url, box_size=size)

    return StreamingResponse(
        io.BytesIO(png_bytes),
        media_type="image/png",
        headers={
            "Cache-Control": "public, max-age=3600",
            "X-QR-URL": url,
        },
    )

File: server/internal/handler/qr.go
package handler

import (
	"net/http"
	"os"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"
	qrcode "github.com/skip2/go-qrcode"
)

const defaultFrontendURL = "https://bulpan.example.com"

// QRHandler serves the GET /api/qr endpoint.
type QRHandler struct{}

// NewQRHandler creates a new QRHandler.
func NewQRHandler() *QRHandler {
	return &QRHandler{}
}

// getMapURL builds the map page URL, optionally with ?demo=true for GPS fallback.
func getMapURL(demo bool) string {
	base := os.Getenv("FRONTEND_URL")
	if base == "" {
		base = defaultFrontendURL
	}
	base = strings.TrimRight(base, "/")
	if demo {
		return base + "?demo=true"
	}
	return base
}

// Handle generates a QR code PNG image that links to the frontend map page.
//
//	GET /api/qr?demo=false&size=10
//	Response: PNG image bytes
//	Headers: Content-Type: image/png, Cache-Control: public, max-age=3600, X-QR-URL: <url>
func (h *QRHandler) Handle(c *gin.Context) {
	// Parse demo parameter (default: false)
	demoStr := c.DefaultQuery("demo", "false")
	demo := demoStr == "true" || demoStr == "1"

	// Parse size parameter (default: 10, range: 4-40)
	sizeStr := c.DefaultQuery("size", "10")
	size, err := strconv.Atoi(sizeStr)
	if err != nil || size < 4 || size > 40 {
		size = 10
	}

	url := getMapURL(demo)

	// Generate QR code PNG bytes
	// go-qrcode doesn't support box_size directly; we compute pixel size.
	// QR version auto-selects. For the URL length we use, version ~2 (25 modules).
	// Total pixels = (modules + 2*border) * box_size
	// We use RecoveryLevel Medium to match Python's ERROR_CORRECT_M.
	// Border of 2 modules (matching Python), default go-qrcode border is 4.
	// go-qrcode uses a fixed border, so we set the total image size.
	// Approximate: version 2 = 25 modules, border=2 => 29 * box_size
	// We'll use -size (negative) for auto-size with the library, but go-qrcode
	// takes a positive pixel size. We'll compute a reasonable size.
	//
	// Actually, go-qrcode.Encode takes total image pixel width.
	// Python: image size = (modules + 2*border) * box_size
	// With auto version for a short URL, typically version 2-3 (25-29 modules).
	// We approximate total pixels similarly.
	modules := 29 // reasonable default for short URLs (version 3)
	border := 2
	totalPixels := (modules + 2*border) * size

	pngBytes, err := qrcode.Encode(url, qrcode.Medium, totalPixels)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to generate QR code"})
		return
	}

	c.Header("Cache-Control", "public, max-age=3600")
	c.Header("X-QR-URL", url)
	c.Data(http.StatusOK, "image/png", pngBytes)
}

// Register adds the QR endpoint to the given Gin router group.
func (h *QRHandler) Register(r gin.IRouter) {
	r.GET("/api/qr", h.Handle)
}
